libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,183 @@
+#include <stdbool.h>
+
+#include "adc.h"
+#include "log.h"
+#include "stm32f0xx.h"
+
+#define TS_CAL1 0x1FFFF7b8
+#define TS_CAL2 0x1FFFF7c2
+#define VREFINT_CAL 0x1FFFF7ba
+
+typedef struct ADCInterrupt {
+  ADCCallback callback;
+  void *context;
+  uint16_t reading;
+} ADCInterrupt;
+
+typedef struct ADCStatus {
+  uint32_t sequence;
+  bool continuous;
+} ADCStatus;
+
+static ADCInterrupt s_adc_interrupts[NUM_ADC_CHANNEL];
+static ADCStatus s_adc_status;
+
+static uint16_t prv_get_temp(uint16_t reading) {
+  uint16_t ts_cal1 = *(uint16_t*)TS_CAL1;
+  uint16_t ts_cal2 = *(uint16_t*)TS_CAL2;
+
+  reading = ((110 - 30) * (reading - ts_cal1)) / (ts_cal2 - ts_cal1) + 30;
+
+  return reading + 273;
+}
+
+static uint16_t prv_get_vdda(uint16_t reading) {
```

Comment:
Please document where you got this formula and the temperature calibration formula.